connectivity/mqtt/client_common.c

```diff
@@ -0,0 +1,613 @@
+#endif
+
+#ifdef WITH_SOCKS
+  cfg->socks_config = (mosq_socks_config_t *)malloc(sizeof(mosq_socks_config_t));
+  if (cfg->socks_config == NULL) {
+    log_error(mqtt_common_logger_id, "[%s:%d:%s]\n", __func__, __LINE__, "SC_TA_OOM");
+    ret = SC_MQTT_OOM;
+    goto oom_err;
+  }
+  memset(cfg->socks_config, 0, sizeof(mosq_socks_config_t));
+#endif
+
+  if ((client_type == client_pub) || (client_type == client_duplex)) {
+    cfg->pub_config = (mosq_pub_config_t *)malloc(sizeof(mosq_pub_config_t));
+    if (cfg->pub_config == NULL) {
+      log_error(mqtt_common_logger_id, "[%s:%d:%s]\n", __func__, __LINE__, "SC_TA_OOM");
+      ret = SC_MQTT_OOM;
+      goto oom_err;
+    }
+    cfg->pub_config->first_publish = true;
+    memset(cfg->pub_config, 0, sizeof(mosq_pub_config_t));
+  }
+  if ((client_type == client_sub) || (client_type == client_duplex)) {
+    cfg->sub_config = (mosq_sub_config_t *)malloc(sizeof(mosq_sub_config_t));
+    if (cfg->sub_config == NULL) {
+      log_error(mqtt_common_logger_id, "[%s:%d:%s]\n", __func__, __LINE__, "SC_TA_OOM");
+      ret = SC_MQTT_OOM;
+      goto oom_err;
+    }
+    memset(cfg->sub_config, 0, sizeof(mosq_sub_config_t));
+  }
+
+  cfg->general_config->port = -1;
+  cfg->general_config->qos = 1;
+  cfg->general_config->max_inflight = 20;
+  cfg->general_config->keepalive = 60;
+  cfg->general_config->clean_session = true;
+  cfg->general_config->protocol_version = MQTT_PROTOCOL_V311;
+  cfg->general_config->client_type = client_type;
+
+  return SC_OK;
+
+oom_err:
+  free(cfg->general_config);
+  free(cfg->property_config);
+  free(cfg->pub_config);
+  free(cfg->sub_config);
+#ifdef WITH_TLS
+  free(cfg->tls_config);
+#endif
+
+#ifdef WITH_SOCKS
+  free(cfg->socks_config);
+#endif
+  return ret;
+}
+
+status_t mosq_config_free(mosq_config_t *cfg) {
+  if (cfg == NULL) {
+    log_error(mqtt_common_logger_id, "[%s:%d:%s]\n", __func__, __LINE__, "SC_TA_NULL");
+    return SC_MQTT_NULL;
+  }
+
+  if (cfg->general_config->client_type == client_pub || cfg->general_config->client_type == client_duplex) {
+    free(cfg->pub_config->message);
+    free(cfg->pub_config->topic);
+    free(cfg->pub_config->response_topic);
+  }
+  if ((cfg->general_config->client_type == client_sub) || (cfg->general_config->client_type == client_duplex)) {
+    if (cfg->sub_config->topics) {
+      for (int i = 0; i < cfg->sub_config->topic_count; i++) {
+        free(cfg->sub_config->topics[i]);
+      }
+      free(cfg->sub_config->topics);
+    }
+    if (cfg->sub_config->unsub_topics) {
+      for (int i = 0; i < cfg->sub_config->unsub_topic_count; i++) {
+        free(cfg->sub_config->unsub_topics[i]);
+      }
+      free(cfg->sub_config->unsub_topics);
+    }
+  }
+  free(cfg->general_config->id);
+  free(cfg->general_config->host);
+  free(cfg->general_config->username);
+  free(cfg->general_config->password);
+  free(cfg->general_config->will_topic);
+  free(cfg->general_config->will_payload);
+
+#ifdef WITH_TLS
```

Comment:
The more `ifdef`, the more conditional builds required. Please do check the necessity for these `ifdef`. It might cause the conditions of non-functional code.

Reply:
Now all the code relative to the object in `ifdef` are removed. Only code for initialization and free remain.
There are only to functionality being protected by `ifdef`. They are `TLS` and `SOCKS` respectively. I think they both are important in our scenario, and I will implement them after finishing implementing therequest handlers.